logicalplan/projection_pushdown.go

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package logicalplan
+
+import (
+	"github.com/thanos-io/promql-engine/query"
+
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/util/annotations"
+)
+
+type ProjectionPushdown struct {
+	SeriesHashLabel string
+}
+
+func (p ProjectionPushdown) Optimize(plan Node, _ *query.Options) (Node, annotations.Annotations) {
+	p.pushProjection(&plan, nil, false)
+	return plan, nil
+}
+
+// pushProjection recursively traverses the tree and pushes projection information down.
+func (p ProjectionPushdown) pushProjection(node *Node, requiredLabels map[string]struct{}, isWithout bool) {
```

Comment:
Could this method be refactored? Maybe independent methods for each type of node types similar to how it's in execution.go?

Reply:
Were you referring to adding another method in the logical node interface? https://github.com/thanos-io/promql-engine/blob/main/logicalplan/logical_nodes.go

I feel the logic of `pushProjection` is kind of specific to this optimizer. I am not sure if it is worth it as it seems not reusable in other places.